crates/scarb-api/src/metadata_command.rs

```diff
@@ -0,0 +1,247 @@
+use crate::{
+    output_transform::{OutputTransform, PassByPrint},
+    ScarbCommand, ScarbCommandError,
+};
+use scarb_metadata::{Metadata, MetadataCommandError, VersionPin};
+use serde_json::Value;
+use std::{borrow::Cow, io::BufRead};
+
+impl From<ScarbCommandError> for MetadataCommandError {
+    fn from(value: ScarbCommandError) -> Self {
+        match value {
+            ScarbCommandError::Io(io) => MetadataCommandError::Io(io),
+            ScarbCommandError::ScarbError { stdout, stderr } => {
+                MetadataCommandError::ScarbError { stdout, stderr }
+            }
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct ScarbErrorPrettyPrint;
+
+fn pretty_print<'a>(out: &'a [u8]) -> Cow<'a, [u8]> {
+    let message = serde_json::from_slice::<'a, Value>(out).map_or_else(
+        |_| String::from_utf8_lossy(out).to_string(),
+        |value| match value["message"].as_str() {
+            Some(message) if value["type"] == "error" => message.to_owned(),
+            _ => serde_json::to_string_pretty(&value).unwrap(),
+        },
+    );
+
+    let mut result = Vec::new();
+
+    result.extend_from_slice(b"\nScarb returned error:\n");
+
+    for line in message.lines() {
+        result.extend(b"Scarb error:    ");
+        result.extend(line.as_bytes());
+        result.extend(b"\n");
+    }
+
+    result.extend(b"\n");
+
+    result.into()
+}
+
+impl OutputTransform for ScarbErrorPrettyPrint {
+    fn transform_stderr(stderr: &[u8]) -> Cow<'_, [u8]> {
+        pretty_print(stderr)
+    }
+    fn transform_stdout(stdout: &[u8]) -> Cow<'_, [u8]> {
+        pretty_print(stdout)
+    }
+}
+
+pub struct ScarbMetadataCommand<Stdout = PassByPrint, Stderr = PassByPrint> {
+    inner: ScarbCommand<Stdout, Stderr>,
```

Comment:
I think this is not a good idea to try to emulate subtyping here. Better idea would be to represent `ScarbCommand` as enum (nicely documents available options) and name a strcut containing shared arguments `ScarbCommandSharedArgs` or something similar.